machineconfiguration/v1alpha1/types_machineosbuild.go

```diff
@@ -0,0 +1,124 @@
+	// configGeneration tracks which version of MachineOSConfig this build is based off of
+	// +kubebuilder:validation:Required
+	ConfigGeneration int64 `json:"configGeneration"`
+	// conditions are state related conditions for the build. Valid types are:
+	// Prepared, Building, Failed, Interrupted, Restarted, and Ready
+	// once a Build is marked as Failed, no future conditions can be set. This is enforced by the MCO.
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+	// buildName is the build object reference associated with this MachineOSBuildObject
+	BuildName string `json:"buildName"`
+	// buildStart describes when the build started.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="buildStart is immutable once set"
+	// +kubebuilder:validation:Required
+	BuildStart *metav1.Time `json:"buildStart"`
+	// buildEnd describes when the build ended.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="buildEnd is immutable once set"
+	//+optional
+	BuildEnd *metav1.Time `json:"buildEnd"`
+	// desiredConfig is the desired config we want to build an image for.
+	// +kubebuilder:validation:Required
+	DesiredConfig RenderedMachineConfigReference `json:"desiredConfig"`
+	// finalImagePullSpec describes the fully qualified pullspec produced by this build that the final image can be. Must be in sha format.
+	// +kubebuilder:validation:XValidation:rule=`(self.split('@').size() == 2 && self.split('@')[1].matches('^sha256:[a-f0-9]{64}$'))`,message="the OCI Image reference must end with a valid '@sha256:<digest>' suffix, where '<digest>' is 64 characters long"
+	// +optional
+	FinalImagePullspec string `json:"finalImagePullspec"`
+}
+
+// BuildProgess highlights some of the key phases of a build to be tracked in Conditions.
+type BuildProgress string
+
+const (
+	// prepared indicates that the build has finished preparing. A build is prepared
+	// by gathering the build inputs, validating them, and making sure we can do an update as specified.
+	MachineOSBuildPrepared BuildProgress = "Prepared"
+	// building indicates that the build has been kicked off with the specified image builder
+	MachineOSBuilding BuildProgress = "Building"
+	// failed indicates that during the build or preparation process, the build failed.
+	MachineOSBuildFailed BuildProgress = "Failed"
+	// interrupted indicates that the user stopped the build process by modifying part of the build config
+	MachineOSBuildInterrupted BuildProgress = "Interrupted"
+	// restarted indicates that this build has been either interrupted or failed, and now a new
+	// iteration of the process has begun.
+	MachineOSBuildRestarted BuildProgress = "Restarted"
+	// ready indicates that the build has completed and the image is ready to roll out.
+	MachineOSReady BuildProgress = "Ready"
```

Comment:
So this is Complete?

Reply:
yep, I can rename it to complete if that is the norm